### PR TITLE
#Sql tasks. sergey.shestak

### PR DESCRIPTION
**Pull request by sergeyshestak**
https://travis-ci.org/github/sergeyshestak/db-assignments/builds/716164969

### CODE REVIEW

File: task/sql-tasks.js [hunk 1]
```diff
@@ -43,7 +43,16 @@ async function task_1_1(db) {
  *
  */
 async function task_1_2(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT DISTINCT
```
Comment (AlexMokin) [inline task/sql-tasks.js]: why do you need distinct here?

File: task/sql-tasks.js [hunk 2]
```diff
@@ -91,7 +125,17 @@ async function task_1_5(db) {
...
  */
 async function task_1_6(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            Products.ProductName,
+            Categories.CategoryName,
+            Suppliers.CompanyName as "SupplierCompanyName"
+        FROM Products
+        JOIN Categories ON Products.CategoryID=Categories.CategoryID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use **inner join** keyword explicitly

File: task/sql-tasks.js [hunk 3]
```diff
@@ -129,7 +191,14 @@ async function task_1_8(db) {
  *
  */
 async function task_1_9(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            CustomerID,
+            ContactName
+        FROM Customers
+        WHERE ContactName RLIKE '^[F].{2}[n]';
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use like

File: task/sql-tasks.js [hunk 4]
```diff
@@ -140,7 +209,14 @@ async function task_1_9(db) {
  *
  */
 async function task_1_10(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            ProductID,
+            ProductName
+        FROM Products
+        WHERE NOT discontinued='0';
```
Comment (AlexMokin) [inline task/sql-tasks.js]: discontinued is not a string, do not compare it to string

File: task/sql-tasks.js [hunk 5]
```diff
@@ -177,7 +271,15 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
```
Comment (AlexMokin) [inline task/sql-tasks.js]: it's better to write this query like this
notice that from-clause is optional.



```
select
   (select count()...) as 'T'
   (select count()...) as 'TT'
```

File: task/sql-tasks.js [hunk 6]
```diff
@@ -199,7 +309,24 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            SUM(MONTH(OrderDate) = 1) as "January",
+            SUM(MONTH(OrderDate) = 2) as "February",
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count records, it shows intent better

File: task/sql-tasks.js [hunk 7]
```diff
@@ -247,7 +399,19 @@ async function task_1_18(db) {
...
 async function task_1_19(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            Customers.CustomerID,
+            Customers.CompanyName,
+            SUM(UnitPrice * Quantity) as "TotalOrdersAmount, $" 
+        FROM Customers
+        JOIN Orders ON Customers.CustomerID=Orders.CustomerID
+        JOIN OrderDetails ON Orders.OrderID=OrderDetails.OrderID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use **inner join** keyword explicitly

File: task/sql-tasks.js [hunk 8]
```diff
@@ -259,7 +423,19 @@ async function task_1_19(db) {
...
 async function task_1_20(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            SUM(UnitPrice * Quantity) as "Amount, $",
+            CONCAT(Employees.FirstName, ' ', Employees.LastName) as "Employee Full Name",
+            Orders.EmployeeID
+        FROM Orders
+        JOIN Employees ON Orders.EmployeeID=Employees.EmployeeID
+        JOIN OrderDetails ON Orders.OrderID=OrderDetails.OrderID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use **inner join** keyword explicitly

File: task/sql-tasks.js [hunk 9]
```diff
@@ -280,7 +465,25 @@ async function task_1_21(db) {
...
 async function task_1_22(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT DISTINCT
+            Customers.CompanyName,
+            Products.ProductName,
+            OrderDetails.UnitPrice as "PricePerItem"
+        FROM Customers
+        JOIN Orders ON Customers.CustomerID=Orders.CustomerID
+        JOIN OrderDetails ON Orders.OrderID=OrderDetails.OrderID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use **inner join** keyword explicitly

File: task/sql-tasks.js [hunk 10]
```diff
@@ -280,7 +465,25 @@ async function task_1_21(db) {
...
+        JOIN Products ON OrderDetails.ProductID=Products.ProductID
+        WHERE OrderDetails.UnitPrice = (SELECT
+                    MAX(OD.UnitPrice)
+                FROM Customers as C
+                JOIN Orders as O ON O.CustomerID=C.CustomerID
+                JOIN OrderDetails as OD ON OD.OrderID=O.OrderID
+                WHERE Customers.CompanyName=C.CompanyName
```
Comment (AlexMokin) [inline task/sql-tasks.js]: company name doesn't uniquely identify customer. also this column doesn't have an index so it's also slow.

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js [hunk 11]
```diff
@@ -129,7 +191,14 @@ async function task_1_8(db) {
  *
  */
 async function task_1_9(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            CustomerID,
+            ContactName
+        FROM Customers
+        WHERE ContactName LIKE 'F%__n%';
```
Comment (AlexMokin) [inline task/sql-tasks.js]: **n** must be exactly 4th character

File: task/sql-tasks.js [hunk 12]
```diff
@@ -177,7 +271,16 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+			(SELECT COUNT(*)
+			FROM Products) as TotalOfCurrentProducts,
+            (SELECT COUNT(CASE
+                WHEN Discontinued=1 THEN true
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use where clause in  this sub-query to filter records

File: task/sql-tasks.js [hunk 13]
```diff
@@ -177,7 +271,16 @@ async function task_1_12(db) {
...
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+			(SELECT COUNT(*)
+			FROM Products) as TotalOfCurrentProducts,
+            (SELECT COUNT(CASE
+                WHEN Discontinued=1 THEN true
+                END)) as TotalOfDiscontinuedProducts
+        FROM Products;
```
Comment (AlexMokin) [inline task/sql-tasks.js]: from clause is optional

**Comment by AlexMokin** [comment]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by IlyaMokin** [comment]
Thanks for the efforts! The PR is accepted!

**Pull request closed without merging**